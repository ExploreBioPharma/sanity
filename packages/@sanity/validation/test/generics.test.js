const {Rule} = require('../src')

describe('generics', () => {
  test('should be able to construct an empty rule', () => {
    expect(() => new Rule()).not.toThrow(Error)
    expect(new Rule()).toBeInstanceOf(Rule)
  })

  test('should be able to construct a new typed rule', () => {
    expect(Rule.string()).toBeInstanceOf(Rule)
  })

  test('clones rule when changing generics', () => {
    const baseRule = new Rule()
    const typed = baseRule.type('string')
    expect(baseRule).not.toBe(typed)
  })

  test('clones rule when changing type-specific rules', () => {
    const baseRule = Rule.string()
    const specific = baseRule.min(5)
    expect(baseRule).not.toBe(specific)
  })

  test('returns arrays of errors/warnings by default', () => {
    expect(Rule.string().validate(123)).toMatchSnapshot()
  })

  test('throws validation errors if told to', () => {
    expect(() => Rule.string().validate(123, {throwOnError: true})).toThrowErrorMatchingSnapshot()
  })

  test('can override earlier defined rules', () => {
    expect(
      Rule.string()
        .max(1)
        .max(2)
        .validate('hei')
    ).toMatchSnapshot()
  })

  test('can demote errors to warnings', () => {
    const result = Rule.string()
      .warning()
      .validate(123)

    expect(result).toMatchSnapshot()
  })

  test('can customize error messages', () => {
    const result = Rule.string()
      .error('Dude it needs to be a string')
      .validate(123)

    expect(result).toMatchSnapshot()
  })

  test('can customize warning messages', () => {
    const result = Rule.string()
      .warning('Dude it should probably be a string')
      .validate(123)

    expect(result).toMatchSnapshot()
  })

  test('can merge rules', () => {
    const rule = new Rule().required()
    const stringRule = Rule.string().min(5)
    const newRule = rule.merge(stringRule)

    expect(rule).not.toBe(stringRule)
    expect(newRule).not.toBe(stringRule)
    expect(rule).not.toBe(newRule)

    const result = newRule.validate('Hei')
    expect(result).toMatchSnapshot()
  })
})
