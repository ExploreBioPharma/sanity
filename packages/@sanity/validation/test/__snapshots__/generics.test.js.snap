// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generics can customize error messages 1`] = `
Object {
  "errors": Array [
    [ValidationError: Dude it needs to be a string],
  ],
  "warnings": Array [],
}
`;

exports[`generics can customize warning messages 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [
    [ValidationError: Dude it should probably be a string],
  ],
}
`;

exports[`generics can demote errors to warnings 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [
    [ValidationError: Expected type "String", got "Number"],
  ],
}
`;

exports[`generics can merge rules 1`] = `
Object {
  "errors": Array [
    [ValidationError: String must be at least 5 characters long],
  ],
  "warnings": Array [],
}
`;

exports[`generics can override earlier defined rules 1`] = `
Object {
  "errors": Array [
    [ValidationError: String must be at most 2 characters long],
  ],
  "warnings": Array [],
}
`;

exports[`generics returns arrays of errors/warnings by default 1`] = `
Object {
  "errors": Array [
    [ValidationError: Expected type "String", got "Number"],
  ],
  "warnings": Array [],
}
`;

exports[`generics throws validation errors if told to 1`] = `"Expected type \\"String\\", got \\"Number\\""`;
